From fff9c33174c1a694624cb8a12fe70c99dc7d22f0 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 10 Nov 2022 11:49:03 +0100
Subject: [PATCH 12/65] env: mmc: select GPT env partition by type guid

Since commit c0364ce1c695 ("doc/README.gpt: define partition type GUID for
U-Boot environment"), a specific type GUID can be used to indicate
the U-Boot environment partition on the device with GPT partition table.

This patch uses this type GUID to found the env partition as fallback
when the partition name property "u-boot,mmc-env-partition" is not present
in config node or if the indicated partition name is not found.

The mmc_offset_try_partition() function is reused, it selects the first
partition with the correct type GUID when the parameter 'str' is NULL.

[Backport of commit 80105d8fd525 ("env: mmc: select GPT env partition by
 type guid")]

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git e6f9449c25deebbc0b95beb1e52e1009aafc75d0 env: mmc: select GPT env partition by type guid]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-by: Patrice Chotard <patrice.chotard@foss.st.com>
Change-Id: I0a9d81be69475c89cac4f007332fc368197d2868
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/284267
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 env/mmc.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/env/mmc.c b/env/mmc.c
index 1eddba3b0f8..f819c954a28 100644
--- a/env/mmc.c
+++ b/env/mmc.c
@@ -191,6 +191,13 @@ static inline s64 mmc_offset(struct mmc *mmc, int copy)
 			return val;
 	}
 
+	/* try the GPT partition with "U-Boot ENV" TYPE GUID */
+	if (IS_ENABLED(CONFIG_PARTITION_TYPE_GUID)) {
+		err = mmc_offset_try_partition(NULL, copy, &val);
+		if (!err)
+			return val;
+	}
+
 	defvalue = ENV_MMC_OFFSET;
 	propname = dt_prop.offset;
 
-- 
2.51.0

